README.md

```diff
@@ -48,33 +65,470 @@ cache read/white process, etc. Command handler can be sync and publish the
+  }
+}
+```
+
+### Middleware
+
+**Middleware** is a layer where commands and events could be intercepted before
+the moment they reach listeners. It means you can modify/cancel/extend the
+executed command in **Command Middleware**, or do appropriate operation in
+**Event Middleware** before the published event is received by its reactions.
+It's handy for logging, crash reporting, aborting particular commands or
+events, etc.
+
+```swift
+// Command middleware
+struct ErrorCommandMiddleware: CommandMiddleware {
+
+  func intercept(command: AnyCommand, execute: Execute, next: Execute) throws {
+    do {
+      // Don't forget to call `next` to invoke the next function in the chain.
+      try next(command)
+    } catch {
+      print("Command failed with error -> \(command)")
+      throw error
+    }
+  }
+}
+
+Engine.sharedInstance.pipeCommands(through: [ErrorCommandMiddleware()])
+
+// Event middleware
+struct LogEventMiddleware: EventMiddleware {
+
+  // Don't forget to call `next` to invoke the next function in the chain.
+  func intercept(event: AnyEvent, publish: Publish, next: Publish) throws {
+    print("Event published -> \(event)")
+    try next(event)
+  }
+}
+
+Engine.sharedInstance.pipeEvents(through: [LogEventMiddleware()])
+```
+
+**Note** that it's necessary to call `next` to invoke the next function in the
+chain while building your custom middleware.
+
+`AnyCommand` and `AnyEvent` are special protocols that every `Command` or
+`Event` conform to. They are used mostly in [middleware](#middleware) to
+workaround restrictions of working with Swift generic protocols that have
+`associatedtype`.
+
+## Engine
+
+**Engine** is the main entry point for **Aftermath** configuration:
+
+- Register command handlers:
+
+```swift
+Engine.sharedInstance.use(BookListCommandHandler())
+```
+
+- Add command and event middleware:
+
+```swift
+// Commands
+Engine.sharedInstance.pipeCommands(through: [LogCommandMiddleware(), ErrorCommandMiddleware()])
+// Events
+Engine.sharedInstance.pipeEvents(through: [LogEventMiddleware(), ErrorEventMiddleware()])
+```
+
+- Set global error handler to catch all unexpected errors and framework
+warnings:
+
+```swift
+struct EngineErrorHandler: ErrorHandler {
+
+  func handleError(error: ErrorType) {
+    if let error = error as? Error {
+      print("Engine error -> \(error)")
+    } else if let warning = error as? Warning {
+      print("Engine warning -> \(warning)")
+    } else {
+      print("Unknown error -> \(error)")
+    }
+  }
+}
+
+Engine.sharedInstance.errorHandler = EngineErrorHandler()
+```
+
+- Dispose all registered command handlers and event listeners (reactions):
+
+```swift
+Engine.sharedInstance.invalidate()
+```
+
+## Life hacks
+
+### Stories
+
+Naming is hard. I doesn't seem right to have names like `BookListCommand`,
+`BookListCommandHandler` and `BookListWhatever` in the project. Here comes the
+idea of stories that group related types and make the flow more concrete:
```

Comment:
From: 
> Naming is hard. I doesn't seem right to have names like `BookListCommand`,
`BookListCommandHandler` and `BookListWhatever` in the project. Here comes the
idea of stories that group related types and make the flow more concrete:

To:
> Naming is hard. It doesn't feel right to have names like `BookListCommand`,
`BookListCommandHandler` and `BookListWhatever`, does it? If you agree, then you can work
around this issue by introducing a new idea into the mix.
You can group all related types into stories, which makes the flow more concrete.